sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,612 @@
+      }
+      builder.addIfNotNull(DisplayData.item("maxReadTime", maxReadTime()));
+    }
+
+  }
+
+  /**
+   * Checkpoint for an unbounded MQTT source. Consists of the MQTT messages waiting to be
+   * acknowledged and oldest pending message timestamp.
+   */
+  @DefaultCoder(AvroCoder.class)
+  private static class MqttCheckpointMark implements UnboundedSource.CheckpointMark {
+
+    private MqttClient client;
+
+    private final List<MqttMessageWithTimestamp> messages = new ArrayList<>();
+    private Instant oldestPendingTimestamp = BoundedWindow.TIMESTAMP_MIN_VALUE;
+
+    public MqttCheckpointMark() {
+    }
+
+    public void setClient(MqttClient client) {
+      this.client = client;
+    }
+
+    public void add(MqttMessageWithTimestamp message) {
+      if (message.getTimestamp().isBefore(oldestPendingTimestamp)) {
+        oldestPendingTimestamp = message.getTimestamp();
+      }
+      messages.add(message);
+    }
+
+    @Override
+    public void finalizeCheckpoint() {
+      for (MqttMessageWithTimestamp message : messages) {
+        try {
+          client.messageArrivedComplete(message.getMessage().getId(),
+              message.getMessage().getQos());
+          if (message.getTimestamp().isAfter(oldestPendingTimestamp)) {
+            oldestPendingTimestamp = message.getTimestamp();
+          }
+        } catch (Exception e) {
+          LOGGER.warn("Can't ack message {} with QoS {}", message.getMessage().getId(),
+              message.getMessage().getQos());
+        }
+      }
+      messages.clear();
+    }
+
+  }
+
+  private static class UnboundedMqttSource
+      extends UnboundedSource<byte[], MqttCheckpointMark> {
+
+    private final Read spec;
+
+    public UnboundedMqttSource(Read spec) {
+      this.spec = spec;
+    }
+
+    @Override
+    public UnboundedReader<byte[]> createReader(PipelineOptions options,
+                                        MqttCheckpointMark checkpointMark) {
+      return new UnboundedMqttReader(this, checkpointMark);
+    }
+
+    @Override
+    public List<UnboundedMqttSource> generateInitialSplits(int desiredNumSplits,
+                                                           PipelineOptions options) {
+      // MQTT is based on a pub/sub pattern
+      // so, if we create several subscribers on the same topic, they all will receive the same
+      // message, resulting to duplicate messages in the PCollection.
+      // So, for MQTT, we limit to number of split ot 1 (unique source).
+      return Collections.singletonList(new UnboundedMqttSource(spec));
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+    }
+
+    @Override
+    public Coder<MqttCheckpointMark> getCheckpointMarkCoder() {
+      return AvroCoder.of(MqttCheckpointMark.class);
```

Comment:
Since you're specifying the coder here, there's no need to specify `@DefaultCoder` on the class itself. However, my comment about encodability of the inner `MqttClient` stands.